### PR TITLE
[vLLM IR] Add IR op testing and benchmarking infrastructure

### PR DESCRIPTION
**Pull request by gmagogsfm**
Added infrastructure for each vLLM IR Op to define its example input generator and tolerance levels. This enables

- Automatic accuracy testing across all providers without developers effort
- Uniform benchmarking

Example benchmarking output
```
============================================================
Timestamp:      2026-04-17 17:01:01
Git commit:     37464b68b
vLLM:           0.17.1rc1.dev349+g6792117d7.d20260327 (git sha: 6792117d7, date: 20260327)
PyTorch:        2.11.0+cu130
CUDA runtime:   13.2.78
Triton:         3.6.0
CUTLASS:        4.4.1
Helion:         0.2.4.dev3+g34d4a8049
GPU:            GPU 0: NVIDIA H100 80GB HBM3
Bench mode:     cuda_graph
Warmup:         25 ms (ignored)
Repetitions:    100 replays
============================================================

vllm_c vs native (84/84 faster, 0/84 slower):
  geomean speedup: 6.48x
  best:            15.85x  (tokens=2048_hidden=4096_float16)
  worst:           2.62x  (tokens=32768_hidden=16384_float32)

Results saved to: /tmp/vllm_ir_bench_20260417_170101
```

### CODE REVIEW

File: tests/ir/test_ir_ops.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def _output_dtype(out) -> torch.dtype:
+    if isinstance(out, torch.Tensor):
+        return out.dtype
+    if isinstance(out, (tuple, list)):
+        for o in out:
+            if isinstance(o, torch.Tensor):
+                return o.dtype
+    raise TypeError(f"Cannot infer dtype from output: {type(out)}")
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: Good catch. fixed

File: tests/ir/test_ir_ops.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+
+@pytest.fixture(autouse=True)
+def _set_default_device():
+    prev = torch.get_set_default_device()
+    torch.set_set_default_device(current_platform.device_type)
+    yield
+    torch.set_set_default_device(prev)
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: done

File: tests/ir/test_ir_ops.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        for provider, impl in non_native:
+            if not impl.supports_args(*args):
+                continue
+
+            out_impl = impl.impl_fn(*_clone_args(args))
+            dtype = _output_dtype(out_impl)
+            tol = op.get_tolerance(dtype)
+            try:
+                torch.testing.assert_close(out_impl, out_native, **tol)
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: done

File: vllm/ir/ops/layernorm.py [hunk 4]
```diff
@@ -19,3 +19,27 @@ def rms_norm(
     if weight is not None:
         x = x.to(weight.dtype) * weight
     return x.to(orig_dtype)
+
+
+@rms_norm.register_input_generator
+def _rms_norm_input_generator():
+    num_tokens = [1, 4, 32, 128, 512, 2048, 32768]
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply vllm/ir/ops/layernorm.py]: added seed setting to benchmarking and testing logic, it should not be in input generator

File: benchmarks/kernels/ir/bench_ir_ops.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+    with contextlib.suppress(subprocess.CalledProcessError, FileNotFoundError):
+        git_sha = (
+            subprocess.check_output(
+                ["git", "rev-parse", "--short", "HEAD"], stderr=subprocess.DEVNULL
+            )
+            .decode()
+            .strip()
+        )
+
+    gpu = env.nvidia_gpu_models or "unknown"
```
Comment (ProExpertProg) [inline benchmarks/kernels/ir/bench_ir_ops.py]: @tjtanaa could your team help making this portable to other platforms as well?

File: vllm/ir/op.py [hunk 6]
```diff
@@ -13,6 +13,39 @@
 from vllm.logger import init_logger
 from vllm.logging_utils import lazy, tensors_str_no_data
 
+InputGenerator = Callable[[], dict[str, tuple[Any, ...]]]
```
Comment (ProExpertProg) [inline vllm/ir/op.py]: Let's put this into `tolerances.py` or `testing.py` (same folder)?

File: benchmarks/kernels/ir/bench_ir_ops.py (hunk 5 above)
Comment (gmagogsfm) [reply benchmarks/kernels/ir/bench_ir_ops.py]: Please let me know what other platforms to cover and the canonical methods to get their name, happy to incorporate here

**Review by ProExpertProg — COMMENTED** [review]
Great work! Implementation looks solid. I also really like the concept of tolerances accompanying the ops, which allows the "accuracy" of the op (tolerance) to accompany its semantic definition.

I'm not that sure that input generation should live in `vllm/ir`. I worry that (1) it doesn't make sense for the IR to decide what shapes matter, (2) it couples the benchmarking and testing shapes and (3) if people want to adjust the number of shapes in the test, they'll have to modify the `ir` folder.

Instead, what if each op just registers a mapping from `num_tokens, hidden_size` pairs and returns the corresponding inputs. Tests and benchmarks can then create those pairs however they want. This also allows us to reuse a base list of size pairs we can have for popular models and deployments.

**Comment by ProExpertProg** [comment]
Also, once we have the option to compile the native implementation (#38775), will that be used as the baseline in benchmarks?

**Comment by ProExpertProg** [comment]
Also, once we have the option to compile the native implementation (#38775), will that be used as the baseline in benchmarks?

**Comment by gmagogsfm** [comment]
> Also, once we have the option to compile the native implementation (#38775), will that be used as the baseline in benchmarks?

For sure, we should compare against True baseline, ie. whatever vLLM runs by default.

**Comment by gmagogsfm** [comment]
> Great work! Implementation looks solid. I also really like the concept of tolerances accompanying the ops, which allows the "accuracy" of the op (tolerance) to accompany its semantic definition.
> 
> I'm not that sure that input generation should live in `vllm/ir`. I worry that (1) it doesn't make sense for the IR to decide what shapes matter, (2) it couples the benchmarking and testing shapes and (3) if people want to adjust the number of shapes in the test, they'll have to modify the `ir` folder.
> 
> Instead, what if each op just registers a mapping from `num_tokens, hidden_size` pairs and returns the corresponding inputs. Tests and benchmarks can then create those pairs however they want. This also allows us to reuse a base list of size pairs we can have for popular models and deployments.

Redesigned input generator, please take another look

File: vllm/ir/op.py (hunk 6 above)
Comment (gmagogsfm) [reply vllm/ir/op.py]: Done

File: benchmarks/kernels/ir/bench_ir_ops.py (hunk 5 above)
Comment (ProExpertProg) [reply benchmarks/kernels/ir/bench_ir_ops.py]: AMD is the most important, others would be good too

File: tests/ir/test_ir_ops.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+# TODO(gmagogsfm): When the `maybe_inplace` PR lands, ops marked as
+# inplace=True will mutate inputs. We must also compare mutated inputs
+# (not just outputs) between native and provider for those ops.
+#
+# TODO(gmagogsfm): Support more flexible correctness criteria beyond atol/rtol,
+# e.g. "no more than X% of elements may exceed tolerance". This is useful for
+# ops like rms_norm where reductions cause a tiny fraction of elements to
+# exceed strict bounds at large shapes.
+@pytest.mark.parametrize("op_name", _ops_with_generators())
```
Comment (ProExpertProg) [inline tests/ir/test_ir_ops.py]: I don't think it's good to have a single test for all ops. I think it would be better to have separate tests (likely also files) for each op. Each test should be parametrized on the shapes and (supported) providers, and then in the test we can call utilities to make writing the test easier.

File: tests/ir/test_ir_ops.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+def test_correctness_vs_native(op_name):
+    torch.manual_seed(0)
+    op = IrOp.registry[op_name]
+    non_native = [
```
Comment (ProExpertProg) [inline tests/ir/test_ir_ops.py]: We should parametrize the test on this

File: tests/ir/test_ir_ops.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        for name, impl in op.impls.items()
+        if name != "native" and impl.supported
+    ]
+    if not non_native:
+        pytest.skip(f"No non-native providers supported for {op_name}")
+
+    inputs = op.generate_example_inputs(SHAPE_CONFIGS[op_name])
+    failures = []
+
+    for case_name, args in inputs.items():
```
Comment (ProExpertProg) [inline tests/ir/test_ir_ops.py]: Also parametrize on this I think

File: tests/ir/shapes.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""
+Shape configurations for IR op correctness tests.
+
+Uses smaller shapes than benchmarks for fast CI turnaround.
+"""
+
+import torch
+
+SHAPE_CONFIGS: dict[str, list[dict]] = {
```
Comment (ProExpertProg) [inline tests/ir/shapes.py]: This can live on the test parameters as test is per-op

File: benchmarks/kernels/ir/shapes.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+realistic production workloads.
+"""
+
+import torch
+
+SHAPE_CONFIGS: dict[str, list[dict]] = {
+    "rms_norm": [
+        {"num_tokens": n, "hidden_size": d, "dtype": dtype}
+        for dtype in [torch.float16, torch.bfloat16, torch.float32]
+        for d in [4096, 7168, 8192, 16384]
```
Comment (ProExpertProg) [inline benchmarks/kernels/ir/shapes.py]: Extract this into `HIDDEN_SIZES` and add comments which model each size belongs to?

File: benchmarks/kernels/ir/bench_ir_ops.py (hunk 5 above)
Comment (gmagogsfm) [reply benchmarks/kernels/ir/bench_ir_ops.py]: > AMD is the most important, others would be good too

Switched to using `current_platform.get_device_name()` to get platform canonical name

**Review by ProExpertProg — COMMENTED** [review]
I think this is basically what I had in mind! The only remaining concern I have is it's a bit too abstracted/automatic; if someone wants to modify a test they'll start specializing the code and soon it'll be very confusing what happens. I think it's better to be _slightly_ more verbose so we're not opaque. I think the best structure would be something like this (not exactly, just a rough idea):

```

HIDDEN_SIZES

@pytest.mark.parametrize("hidden_size", HIDDEN_SIZES)
@pytest.mark.parametrize("num_tokens", NUM_TOKENS)
@pytest.mark.parametrize("dtype", )
@pytest.mark.parametrize("epsilon", [1e-5, 1e-6])
@pytest.mark.parametrize("provider", ir_test_utils.supported_providers())
def test_rms_norm(hidden_size: int, num_tokens: int, dtype: torch.dtype, provider: str):
    args = ir.ops.rms_norm.generate_inputs(hidden_size: int, num_tokens: int, ...)
    if not ir.ops.rms_norm.impls[provider].supports_args(*args):
        pytest.skip()

    # could also use priority instead of direct call
    ref_output = ir.ops.rms_norm.impls["native"].impl_fn(*args)
    output = ir.ops.rms_norm.impls[provider].impl_fn(*args)
    ir_test_utils.allclose(ir.ops.rms_norm, output, ref_output) 
    # maybe ^ are just one util call?

    # Does the torch opcheck, also other checks (making sure impl can be traced with make_fx, making sure supports_args doesn't specialize on batch size (num_tokens), etc.)
    ir_test_utils.opcheck(ir.ops.rms_norm[provider])
```

File: benchmarks/kernels/ir/shapes.py (hunk 11 above)
Comment (gmagogsfm) [reply benchmarks/kernels/ir/shapes.py]: Done

File: tests/ir/shapes.py (hunk 10 above)
Comment (gmagogsfm) [reply tests/ir/shapes.py]: Done

File: tests/ir/test_ir_ops.py (hunk 9 above)
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: Done

File: tests/ir/test_ir_ops.py (hunk 8 above)
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: Done

File: tests/ir/test_ir_ops.py (hunk 7 above)
Comment (gmagogsfm) [reply tests/ir/test_ir_ops.py]: Done

File: benchmarks/kernels/ir/shapes.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""
+Shape configurations for IR op benchmarks.
+"""
+
+import torch
+
+NUM_TOKENS = [1, 2, 4, 16, 64, 256, 1024, 4096, 16384, 32768]
```
Comment (ProExpertProg) [inline benchmarks/kernels/ir/shapes.py]: Does anyone `max_num_batched_tokens>16384`?

File: tests/kernels/ir/test_layernorm.py [hunk 13]
```diff
@@ -5,15 +5,24 @@
...
 from tests.kernels.allclose_default import get_default_rtol
 from vllm import ir
 from vllm.platforms import current_platform
 
 
-def rms_norm_inputs(n_tokens: int, hidden_size: int, dtype: torch.dtype):
-    x = torch.randn(n_tokens, hidden_size, dtype=dtype)
-    weight = torch.rand(hidden_size, dtype=dtype)
-    return x, weight
+def rms_norm_inputs(
```
Comment (ProExpertProg) [inline tests/kernels/ir/test_layernorm.py]: I think we can just inline this method?

File: tests/kernels/ir/test_ir_ops.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ProExpertProg) [inline tests/kernels/ir/test_ir_ops.py:1]: Let's move this to `tests/kernels/`. `tests/ir` is for IR infra tests only

**Review by ProExpertProg — APPROVED** [review]
Looks good, thanks for this work!

One final ask: could we add some IR infra unit tests? Can be pretty narrow

File: tests/kernels/ir/test_layernorm.py (hunk 13 above)
Comment (gmagogsfm) [reply tests/kernels/ir/test_layernorm.py]: Sure, done.

File: tests/kernels/ir/test_ir_ops.py (hunk 14 above)
Comment (gmagogsfm) [reply tests/kernels/ir/test_ir_ops.py:1]: Good call, done.

File: benchmarks/kernels/ir/shapes.py (hunk 12 above)
Comment (gmagogsfm) [reply benchmarks/kernels/ir/shapes.py]: Yes, but quite uncommon. Removed now.

**Comment by gmagogsfm** [comment]
> Looks good, thanks for this work!
> 
> One final ask: could we add some IR infra unit tests? Can be pretty narrow

Added some light weight unit tests. Going forward, all op-specific tests would be good exercise of the benchmarking/testing infra too.

**Review by ProExpertProg — APPROVED** [review]
